Icons: use plain SVG files, load on-demand

leeoniya opened the pull request:
on a quest to #32328. and one strategy to fix https://github.com/grafana/grafana/issues/29478.

this is an attempt to eliminate the monolithic 2.3MB (non-gzipped) Unicons bundle that is loaded for every user and eagerly blocks everything else. 97% of these icons are unused but there is a React component for each one of them (~1,200 components).

the strategy is to use plain svg files, which has some other benefits (such as interop with external tools, ux, etc.). these files are `fetch()`d on demand by the wrapper `Icon` component. this aims to keep the existing API of providing a string icon name and still benefit from lazy-loading each icon only when used. the result is a complete elimination of the icons bundle, faster load times on imperfect connections and a 36s production build time decrease on my machine (269s -> 233s).

**to test this, you will need to unzip `/public/img/icons/icons.zip` into the same directory**. i did not want to commit 1200 svg files as part of this PR and adding the non-react `unicons` dependency results in huge additions to `yarn.lock` and downloads an extra 21.7MB, just so we can pull 633kb of svg files. we should probably use webpack to download them and extract directly or just include and ship them ourselves.

`@iconscout/unicons` stats:

https://packagephobia.com/result?p=%40iconscout%2Funicons
https://npm.broofa.com/?q=@iconscout/unicons

i had to manually convert our own custom icons to svg from js component wrappers (`Icon/assets`, `Icon/custom`), and did a bit of SVG cleanup in the process. to me everything still looks as it should. these icons had to be moved out into `/public/img/icons` so they can be fetched on demand.

~~one drawback is an extra 42 fetch requests when loading the GraphNG test dashboard. we can address in a follow-up (if needed) via pre-caching, using an svg sprite-sheet, etc. however, this is much more desirable than blocking the UI render entirely to wait for a large JS download, decompression and JS parsing.~~ **SOLVED (see below)**. a more simple route to this is to make `<img src="/icons/icon.svg">` elements. but a limitation of this is you lose the ability to style them with css via `fill` and `currentColor`, which unfortunately limits theming options :(. the best of all worlds would be to explicitly import the icons (as React components) so the rest can simply be tree-shaken out, but the by-string lookup API leaves only limited routes to streamline this.

would be curious to see if anyone has additional ideas :)

stats that have changed:

before:

```
Production Build: 269s

JS:   10.3MB
DCL:  1058ms

Throttled to 4G speeds
  DCL:    21.3s
  Load:   22.5s
  Finish: 24.2s
```

after:

```
Production Build: 233s

JS:   8.0MB
DCL:  974ms

Throttled to 4G speeds
  DCL:    16.7s
  Load:   18.0s
  Finish: 19.7s

Extra icon fetchs:
  Requests: 42
  Payload:  27.3 kB
```

bundles before:

![image](https://user-images.githubusercontent.com/43234/112733083-d66fde80-8f0b-11eb-94f1-2eb47d470e7b.png)


bundles after:

![image](https://user-images.githubusercontent.com/43234/112733095-e38ccd80-8f0b-11eb-8788-de7565a5ce02.png)

leeoniya committed:
Icons: use plain SVG files, load on-demand

leeoniya commented:
a more simple route to this is to make `<img src="/icons/icon.svg">` elements. but a drawback to this is you lose the ability to style them with css via `fill`, which unfortunately limits theming options :(

leeoniya commented:
i've re-run the numbers with gzip enabled locally. the gains are much less impressive of course, this shaves about 250kb off the transfer payload and 500ms off the DOMContentLoaded event when throttling to 4G speeds. i'm running these tests on an 8 core machine with 16GB of ram, so the CPU effect on mobile devices can be more significant than on a device that's plugged into power, with effectively unlimited resources.

the best of all worlds would be to explicitly import the icons (as react components) so the rest can simply be tree-shaken out, but the by-string lookup API leaves only with limited options to streamline this.

would be curious to see if anyone has additional ideas :)

hugohaggmark commented on package.json:
```diff
@@ -270,6 +270,7 @@
     "react-dom": "17.0.1",
     "react-grid-layout": "1.2.0",
     "react-highlight-words": "0.16.0",
+    "react-inlinesvg": "^2.2.2",
```
Please use exact versions :)

hugohaggmark commented on package.json:
```diff
@@ -270,6 +270,7 @@
     "react-dom": "17.0.1",
     "react-grid-layout": "1.2.0",
     "react-highlight-words": "0.16.0",
+    "react-inlinesvg": "^2.2.2",
```
Please use exact versions :)

leeoniya replied:
:facepalm:

leeoniya replied:
```suggestion
    "react-inlinesvg": "2.2.2",
```

hugohaggmark reviewed (commented):
Code looks good, not sure what the theming limitations will cause moving forward

leeoniya commented:
> Code looks good, not sure what the theming limitations will cause moving forward

FWIW, i think this has equal theming provisions to the `react-unicons` dependency, which simply spreads props to the root svg element [1]. our own custom icons follow the same pattern [2]. i imagine any deeper icon theming would be impossible anyways with this dependency setup.


[1] https://github.com/Iconscout/react-unicons/blob/master/icons/uil-0-plus.js
[2] https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Icon/assets/Apps.tsx

leeoniya committed:
exact dep versions

torkelo committed:
TestData: Don't add query model for every testdata query scenario by default (#32388)

torkelo committed:
Shortcuts: Fixes panel shortcuts so they always work (#32385)

* Shortcuts: Fixes panel shortcuts so they always work

* Updated snapshot

domasx2 committed:
NgAlerting: loki & cortex have different prom & ruler endpoint prefixes (#32344)

torkelo committed:
ImportDashboard: Fixes backend handler not found error when importing dashboard (#32391)

torkelo committed:
Variables: Fixes error when manually non-matching entering custom value in variable input/picker (#32390)

torkelo committed:
PanelEditor: Fixes issue with data source picker not updating when selecting mixed data source (#32386)

torkelo committed:
DashboardQueryEditor: Run query after selecting source panel (#32383)

bergquist committed:
API: Datasource endpoint should return 400 bad request if id and orgId is invalid (#32392)

* return 400 bad request if id and orgId is invalid

Signed-off-by: bergquist <carl.bergquist@gmail.com>
Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

dependabot[bot] committed:
Chore(deps): Bump github.com/aws/aws-sdk-go from 1.38.2 to 1.38.7 (#32399)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.38.2 to 1.38.7.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.38.2...v1.38.7)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dsotirakis committed:
Update logger message to be more generic (#32403)

kaydelaney committed:
AppRootPage: Fix tests (#32364)

kaydelaney committed:
Table: Fixes so links work for image cells (#32370)

* Panel/Table: Adjust data links logic
Closes #31576

dprokop committed:
Fix inline code being rendered as block in the panel description (#32405)

kaydelaney committed:
Panel/DashList: Fetch folders with View permission (#32406)

Closes #31357

marefr committed:
Postgres/MySQL/MSSQL: Don't return connection failure details to the client (#32408)

For security reasons, log any SQL connection error 
details rather than returning it to the client.

Fixes #26623
Fixes #22000

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

leeoniya commented on yarn.lock:
```diff
@@ -21918,6 +21928,14 @@ react-immutable-proptypes@^2.1.0:
   resolved "https://registry.yarnpkg.com/react-immutable-proptypes/-/react-immutable-proptypes-2.1.0.tgz#023d6f39bb15c97c071e9e60d00d136eac5fa0b4"
   integrity sha1-Aj1vObsVyXwHHp5g0A0TbqxfoLQ=
 
+react-inlinesvg@^2.2.2:
```
```suggestion
react-inlinesvg@2.2.2:
```

ying-jeanne committed:
Docs: /sources/http_api/folder.md (#32407)

mckn committed:
Chore: changed so the PanelChrome story can be changed via controls. (#32105)

torkelo committed:
Variables: Fixes problem with data source variable when default ds is selected (#32384)

wbrowne committed:
remove unnecessary state (#32278)

torkelo committed:
TimeSeries: Fixes issue with time x-axis and hour ticks that stretches across more than one day (#32389)

* TimeSeries: Fixes issue with time x-axis and hour ticks that stretches across more than one day

* Fixed logic

bergquist committed:
API: Return 400 on invalid Annotation requests (#32429)

Signed-off-by: bergquist <carl.bergquist@gmail.com>

bergquist committed:
API: Return 409 on datasource version conflict (#32425)

Signed-off-by: bergquist <carl.bergquist@gmail.com>
Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>

kaydelaney committed:
BarGauge: Fix percentage-based thresholds (#32415)

Closes #31756

torkelo committed:
PanelEdit: Remove right side padding for options pane (#32427)

dsotirakis committed:
Chore: Change from double to single quotes (#32430)

codesome committed:
AlertingNG: Fix dispatcher metrics in notifier (#32434)

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

papagian committed:
[Alerting] Forking alert manager API (#32300)

* Alertmanager lotex ruler

* Apply suggestions from code review

alexanderzobnin committed:
Access-control: use role UID when adding/removing roles (#32438)

torkelo committed:
Graph: Fix setting right y-axis when standard option unit is configured (#32426)

eamonryan committed:
Docs: Clarified Enterprise white-labeling footer text (#32269)

* Clarified white-labeling footer text

* Update docs/sources/enterprise/white-labeling.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* Update docs/sources/enterprise/white-labeling.md

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

* Update docs/sources/enterprise/white-labeling.md

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

* Updated wording

* Updated capitalisation

* Update docs/sources/enterprise/white-labeling.md

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>
Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

oddlittlebird committed:
Docs: Update _index.md (#32447)

oddlittlebird committed:
Edit UI text (#32450)

* UI text edits

* Update index.go

achatterjee-grafana committed:
Updated relref to "Dashboard JSON model:" topic. (#32448)

* Updated relref to "Dashboard JSON model:" topic.

* Update docs/sources/administration/provisioning.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* Update docs/sources/administration/provisioning.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* Update docs/sources/administration/provisioning.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* Added relref (actually).

* Few more changes.

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

achatterjee-grafana committed:
Added "message" to the create dashboard example. (#32452)

oddlittlebird committed:
Edit admin UI text (#32454)

* Update AdminEditOrgPage.tsx

* Update AdminOrgsTable.tsx

* Update AdminSettings.tsx

* Update UserCreatePage.tsx

* Update UserLdapSyncInfo.tsx

* Update UserListAdminPage.tsx

* Update UserOrgs.tsx

* Update UserProfile.tsx

* Update UserSessions.tsx

oddlittlebird committed:
Annotation UI text edits (#32458)

* Update event_editor.html

* Update annotations_srv.ts

* Update editor_ctrl.ts

* Update standardAnnotationSupport.ts

Commit:
Elasticsearch: use application/x-ndjson content type for multisearch requests (#32282)

achatterjee-grafana committed:
Updated docs for issue 32378 (#32462)

* Reorganized content.

* More changes.

* More content reorg and added image links.

leeoniya committed:
implement a pre-primed SVG cache to squash fetch() requests

leeoniya commented:
i've implemented a static SVG cache, which inlines the SVGs of our choosing into the bundle and primes the cache of `react-inlinesvg`. i've had to fork and re-publish this package to expose the global `cacheStore`. that fork is here: https://github.com/grafana/react-inlinesvg/tree/static-cacheStore. i published it to npm as `@leeoniya/react-inlinesvg@2.2.4` until we PR this one-liner upstream or re-publish it under `@grafana` namespace.

this eliminates the additional 42 fetch requests.

ryantxu committed:
load icons from webpack

ryantxu committed:
Merge remote-tracking branch 'origin/master' into icons-svg-lazy-load

ryantxu committed:
add raw-loader dependency directly

ivanahuckova reviewed (commented):
@leeoniya awesome that you are looking into this. I am wondering, do you think there could be done something directly on the `react-unicons` side? So instead of fixing this in Grafana, we could commit to their project? Just curious, cause it would be cleaner on our side. And also, as it is open-source project, we would contribute to their community as well.

leeoniya commented:
@ivanahuckova 

i don't think this an issue with `react-unicons`, it's an issue of not using each icon directly, but instead doing a lookup by string from their entire bundle. this prevents us from tree-shaking all the unused icons. i've been told that we must maintain the existing Icon API of a string icon name, so the current simplified situation is this:

```js
// monolithic unicons import looks like this:
const unicons = {
  'icon1': <react component wrapper for icon1>,
  'icon2': <react component wrapper for icon2>,
  // ...1200 times
}

function getIconComponent(name) {
  return unicons[name];
}
```

since `getIconComponent()` has to be able to load any arbitrary icon by string, the _entire_ unicons bundle must always exist.

if we were able to import each icon component directly, and remove the by-name lookup, then webpack would not include the unused icons. but this not an option.

there is another way to do this with dynamic imports:

```js
function getIconComponent(name) {
  return import("unicons/icons/" + name);
}
```

this probably would work, _but_, webpack then has to potentially compile 1,200 individual modules (one per icon), which would only improve the lazy-loading, but at best do nothing to build times and at worst make them longer.

the route in this PR is a compromise that:
1. maintains the existing by-string icon lookup api
2. lazy loads
3. limits dependency bloat that would result from a direct dependency on the npm `unicons` package
4. improves build times very significantly

i don't know if there is another way to accomplish all four things, but this situation is of our own making and there's not much `react-unicons` can do about it.

i don't fancy the idea of maintaining our own fork of `react-inlinesvg`, so i've opened up an issue in the repo for feedback: https://github.com/gilbarbara/react-inlinesvg/issues/167

leeoniya committed:
use npm-published @leeoniya/react-inlinesvg

leeoniya committed:
converted custom & assets svg files

leeoniya committed:
Merge branch 'master' into icons-svg-lazy-load

ivanahuckova commented:
Thanks so much for walking me trough and explaining the problem in detail. I really appreciate it. 🙏 💙 I see what is the problem now and the solutions and code looks good and makes sense.

leeoniya committed:
cleanup

leeoniya committed:
Merge branch 'master' into icons-svg-lazy-load

torkelo commented:
Great work on this @leeoniya  looking good. Will try to take this for a spin tomorrow

leeoniya committed:
lockfile re-sync

leeoniya commented on package.json:
```diff
@@ -203,6 +205,7 @@
     "@emotion/core": "10.0.27",
     "@grafana/aws-sdk": "0.0.3",
     "@grafana/slate-react": "0.22.9-grafana",
+    "@leeoniya/react-inlinesvg": "2.2.4",
```
i had to expose the global `cacheStore` so i can pre-prime it with the icons we include in the js bundle. it's the first bullet point here: https://github.com/gilbarbara/react-inlinesvg/issues/167

hopefully it gets upstreamed and we can unfork

leeoniya commented:
i'm having some trouble getting the tests to pass. its not clear what was being expected, either:


/public/app/core/components/sidemenu/TopSectionItem.test.tsx:

```js
describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
```

leeoniya committed:
update snapshot

leeoniya committed:
webpack: CopyUniconsPlugin()

leeoniya committed:
slash less

leeoniya committed:
Merge branch 'master' into icons-svg-lazy-load

leeoniya committed:
raw-loader exact version

leeoniya committed:
use lightweight unicons dep

leeoniya commented:
now that the cloudwatch improvements have been merged, this is where we'll be with this PR:

![image](https://user-images.githubusercontent.com/43234/113226094-e877b700-9254-11eb-8e2f-5ccfafe29c7e.png)

leeoniya committed:
remove react-unicons from grafana-ui/package.json

dprokop commented on packages/grafana-ui/src/components/Icon/Icon.tsx:
```diff
@@ -1,14 +1,104 @@
+
+// inlined static cache
+
+import i001 from '!!raw-loader!../../../../../public/img/icons/unicons/bars.svg';
```
Nit: move to iconsCache.ts next to Icon.tsx

torkelo replied:
I moved this to a file named iconBundle.ts and I also changed so this is directly imported from app.ts so that it's not part to any @grafana/ui bundle as the icons are really in the app anyway.

dprokop reviewed (approved):
@leeoniya i think this is a very interesting move and it works great as far as I have tested. This kind of optimizations are always welcome, the impact seems to be signifficant.

The only regression I have noticed is Storybook Icons story not showing icons anymore, as the icons path cannot be resolved. Other than that huge thumbs up.

torkelo committed:
Merge remote-tracking branch 'origin/master' into icons-svg-lazy-load

torkelo committed:
Move pre cached icons to iconBundle

torkelo committed:
Added missing file

torkelo reviewed (approved):
Think this should be good to merge

leeoniya commented:
thanks for the review, guys :)

> Think this should be good to merge

what about dominik's comment?:

> The only regression I have noticed is Storybook Icons story not showing icons anymore, as the icons path cannot be resolved. Other than that huge thumbs up.
